cluster/etcd.go

```diff
@@ -0,0 +1,262 @@
+
+package cluster
+
+import (
+	"log"
+	"strings"
+	"time"
+
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/tools/cache"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/api/rbac/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+)
+
+// SetupEtcCluster performs the setup of the m3 main etcd cluster.
+//
+// This will setup `etcd-operator` by parts, the cluster roles, cluster role bindings, the controller and finally
+// the deployment of the customer resourced named `m3-etc-cluster`.
+//
+// https://github.com/coreos/etcd-operator/
+//
+func SetupEtcCluster() chan error {
+	ch := make(chan error)
+
+	go func() {
+		defer close(ch)
+		clientset, err := k8sClient()
+		if err != nil {
+			ch <- err
+			return
+		}
+		// setup rbac
+		rbac := getEtcdRbacClusterRole()
+		if _, err = clientset.RbacV1beta1().ClusterRoles().Create(rbac); err != nil {
+			ch <- err
+			return
+		}
+		rBinding := getEtcdRbacClusterRoleBinding()
+		if _, err = clientset.RbacV1beta1().ClusterRoleBindings().Create(rBinding); err != nil {
+			ch <- err
+			return
+		}
+		// install etcd operator
+		etcOperator := getEtcdDeployment()
+		if _, err = clientset.AppsV1().Deployments("default").Create(etcOperator); err != nil {
+			ch <- err
+			return
+		}
+
+		// informer factory
+		doneCh := make(chan struct{})
+		factory := informers.NewSharedInformerFactory(clientset, 0)
+
+		etcdOperatorReadyCh := make(chan struct{})
+
+		podInformer := factory.Core().V1().Pods().Informer()
+		podInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				pod := obj.(*corev1.Pod)
+				// monitor for etcd-operator pods
+				if strings.HasPrefix(pod.ObjectMeta.Name, "etcd-operator") {
+					log.Println("etcd-operator pod created:", pod.ObjectMeta.Name)
+					close(etcdOperatorReadyCh)
+					close(doneCh)
+				}
+			},
+		})
+
+		go podInformer.Run(doneCh)
+		// wait for the informer to detect etcd-operator being done
+		<-doneCh
+		<-etcdOperatorReadyCh
+
+		// wait for the deployment to be complete
+
+		// deploy the custom resource definition
+		config := getK8sConfig()
+		//  Create a Dynamic Client to interface with CRDs.
+		dynamicClient, _ := dynamic.NewForConfig(config)
+		etcdclustersResource := schema.GroupVersionResource{
+			Group:    "etcd.database.coreos.com",
+			Version:  "v1beta2",
+			Resource: "etcdclusters",
+		}
+
+		crt := getEtcdCRDDeployment("m3-etcd-cluster")
+
+		// we have no choice but to wait up to a minute for the resource to become available since it's created by
+		// the etcd-operator-controller
+		numberOfTries := 0
+		for {
+			if _, err := dynamicClient.Resource(etcdclustersResource).Namespace("default").Create(crt, metav1.CreateOptions{}); err != nil {
+				log.Println(err)
+				// This should break the loop after 10 attempts
+				if numberOfTries > 30 {
```

Comment:
10 or 30 attempts? is it possible to add a constant for this?

Reply:
30, comment updated